[Edbert Geraldy Cangdinata] iP

Berted opened the pull request:
## TedBot
> "Hello! I'm TedBot ヾ(≧▽≦*)o"

TedBot is a _simple_ chatbot made with Java. It:
* is text-based,
* stores all your tasks,
* is ~~fast~~ SUPER FAST!

To get started, simply:
1. Download the JAR file from [here](https://github.com/Berted/ip/releases) 👈
1. Run TedBot in your preferred terminal 👩‍💻
1. Add your tasks ✍
1. Let it store and handle your tasks for you 👍

Most importantly, it's **FREE**!

Features
- [x] Manages simple tasks
- [x] Manages tasks with deadlines
- [x] Manages tasks with a start and end date
- [ ] A GUI (coming soon!)
- [ ] Reminders for deadlines and events (coming soon!)

---

If you're a Java programmer, you can use this codebase to help you practice Java. For example, here's an excerpt of our code, taken from the `Ui` class:
```java
private void outputString(String out, String header) {
    String[] splitOut = out.split("\n");
    System.out.print(header + " ");
    System.out.println(splitOut[0]);
    for (int i = 1; i < splitOut.length; i++) {
        for (int j = 0; j < header.length() + 1; j++) {
            System.out.print(" ");
        }
        System.out.println(splitOut[i]);
    }
}
```

Berted committed:
Adds Greet, Echo, Exit behaviour

Berted committed:
Duke.java: set class fields to private

Berted committed:
Adds basic adding and listing tasks functionality

Berted committed:
Adds Task class to handle task logic

Berted committed:
Task.java: [cleanup] renamed setCompletion method to setDone

Berted committed:
Duke.java: Added tasks are no longer done by default

Berted committed:
Task.java: Modified status icon to improve readability

Berted committed:
Modify user query to be split into command and arguments

User query is a single string. This makes it difficult to
create commands with custom arguments, such as the `mark` command
from Level-3.

As such, let's divide the user query into two strings,
a "command" string and an "argument" string, split
by whitespace of some length.

Splitting it in this manner keeps the syntax familiar to users
familiar with the terminal, whilst maintaining the ability to
have whitespaces for arguments without dealing with escape
characters.

Berted committed:
Adds mark and unmark task as done functionality

Berted committed:
Add logic to check when a task has been marked or unmarked

Berted committed:
Adds Todo, Deadline, and Event task types

Berted committed:
Created TaskList class to handle task list logic

Berted committed:
Adds semi-automated testing with provided batch script

Berted committed:
Refactor user input parsing to be more general

Berted committed:
Adds DukeException to handle invalid user input

Berted committed:
Fix commit 566be2b, files not staged properly

Berted committed:
TaskList.java: Refactor to use Java Collections' ArrayList

Berted committed:
Adds delete task functionality

Berted committed:
[cleanup] Duke.java: Refactors Exception handling for increased readability

Berted committed:
Adds abstract attribute to Task class

Berted committed:
Adds save functionality

Berted committed:
Modifies Deadline and Event classes to now use LocalDate class

Berted committed:
Merge branch 'branch-Level-7'

Berted committed:
Merge branch 'branch-Level-8'

Berted committed:
Modifies Event and Deadline toStorageString() method for LocalDate

Berted committed:
Adds Gradle support

Berted committed:
Adds UI Class to handle input-output

Berted committed:
Updated EXPECTED.txt to match UI class implementation

Berted committed:
Adds Command classes

Berted committed:
Adds Argument and Parser class to handle parsing user commands

Berted committed:
Organize into packages

Berted committed:
Adds Unit Tests for duke.task and duke.parser

Berted committed:
Add JAR-file build support

Berted committed:
Refactor code to Coding Standard

Berted committed:
Added Checkstyle to comply with SE-EDU coding standard

Berted committed:
Adds Find command

Berted committed:
Add JavaDoc comments

Berted committed:
Merge branch 'branch-Level-9'

Berted committed:
Merge branch 'branch-A-CodingStandard'

Berted committed:
Merge branch 'branch-A-JavaDoc'

Berted committed:
Adds basic GUI

jjtan444 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+     * @param filePath the specified file path parameter.
+     */
+    public Duke(String filePath) {
+        this.ui = new Ui();
```
Might want to remove `this` if it is clear which variable is being referred to.

Berted replied:
Thanks for the suggestion! Resolved in dfffdfb5c287aecd0c382ca6f3eeb96ad1a39af6.

jjtan444 commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    @Override
+    public void exec(TaskList tasks, Ui ui, Storage storage) {
+        StringBuilder reply = new StringBuilder();
```
I like your implementation using `StringBuilder`.

Berted replied:
Thanks!

jjtan444 commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,64 @@
+            if (tasks.getTask(this.unmarkIndex).getIsDone()) {
+                tasks.getTask(this.unmarkIndex).setDone(false);
+                ui.showReply(String.format("Alright, I've marked this task as not done:\n  %s",
+                        tasks.getTask(this.unmarkIndex)));
```
I think the indentation is wrong for this statement and the next block?

jjtan444 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,148 @@
+        Argument[] args = processArgs(command);
+        String commandName = args[0].getName();
+
+        switch (commandName) {
```
I like your implementation of `switch` it makes the code look very clean. 👍

Berted replied:
Thanks!

jjtan444 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,148 @@
+     * @return an array of Arguments.
+     */
+    private static Argument[] processArgs(String command) {
+        return Stream.of(command.trim().split("\\s+/"))
```
I like your use of `Stream` it is very unique. 👍

Berted replied:
Thanks!

jjtan444 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,101 @@
+    public TaskList load() throws DukeException {
+        try {
+            Scanner sc = new Scanner(this.saveFile);
+            TaskList ret = new TaskList();
```
Might want to rename `ret` to provide more clarity.

jjtan444 reviewed (commented):
Overall well done, with some unique implementations and very minor issues.

Wu-Lezheng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+        try {
+            this.ui.showStorageLoadingMessage();
+            this.taskList = storage.load();
+            this.ui.showReply(String.format("Save file loaded. You currently have %d tasks.\n", taskList.getLength()));
```
I like how you show the total number of already stored tasks at first.

Berted replied:
Thanks!

Wu-Lezheng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+     * Starts the Duke chatbot.
+     */
+    public void exec() {
+        boolean exitCalled = false;
```
Maybe it will be clearer to rename "exitCalled" to "shouldExit"?

Berted replied:
Thanks for the suggestion! However, this comment no longer applies since this section of the code has been refactored for GUI. As such, will mark it as resolved.

Wu-Lezheng commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,31 @@
+     * @param storage the specified Storage object.
+     * @throws DukeException if an Exception occurs during its execution.
+     */
+    public abstract void exec(TaskList tasks, Ui ui, Storage storage) throws DukeException;
```
Maybe it will be clearer to rename "exec" to "execute"?

Wu-Lezheng commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,60 @@
+     * @param storage unused for FindCommand.
+     */
+    public void exec(TaskList tasks, Ui ui, Storage storage) {
+        TaskList filteredTasks = tasks.filter((task) -> task.hasSubstring(this.keyword));
```
I like your use of lambda expression here.

Berted replied:
Thanks!

Wu-Lezheng commented on src/main/java/duke/command/UnknownCommand.java:
```diff
@@ -0,0 +1,41 @@
+
+    /**
+     * Throws a DukeException as the user has given an unknown command keyword.
+     * @param tasks the specified TaskList object.
```
It would be better to leave a line between description and @param in the JavaDoc. I notice this problem in other methods as well.

Berted replied:
Thanks for the feedback! Resolved in 3d0220ab8b77dafa39a49a690592363485d70a1f.

Wu-Lezheng commented on src/main/java/duke/parser/Argument.java:
```diff
@@ -0,0 +1,49 @@
+ * This class encapsulates user arguments. Each argument consists
+ * of two components, a name, along with body text.
+ */
+public class Argument {
```
It is interesting to have a class dedicated to arguments.

Wu-Lezheng reviewed (commented):
LGTM. Just need to fix some minor naming and JavaDoc issues.

Berted committed:
Merge branch 'branch-Level-10'

Berted committed:
Add assertions on string parsing methods

Several string parsing methods have unstated assumptions.

Let's add assertions to certain private string parsing methods,
especially those with non-trivial logic.

Using assertions is preferred as it can help improve code
maintainability.

Berted committed:
Duke.java: Refactor program flow to remove explict call to readOutput

Everytime Duke's feedback is required, an explicit call to readOutput
method is required to read the output buffer after running a command.
This design flow causes code duplication as after executing a command,
readOutput will always be called anyways.

Let's modify the program flow such that initialization of Duke and
command execution of Duke handles reading from the output stream
and instead returns the feedback string directly to the method
caller.

This program flow is preferred as to prevent accidental mishap such as
forgetting to read (and subsequently) clear the output stream after
executing a command.

Berted committed:
Fixes style issues on JavaDoc comments

Berted committed:
Storage: Refactors load method to comply with SLAP

The load method of Storage currently also handles parsing the starting
from strings.

Let's pull out this parsing logic to a private method.

This is done to comply with the Single Level of Abstraction Principle.

Berted committed:
Ui: Removes deprecated parameters as TedBot moves to a GUI

Berted committed:
Merge pull request #2 from Berted/branch-A-CodeQuality

Improve Code Quality

Berted committed:
Merge branch 'master' into branch-A-Assertions

Berted committed:
Merge pull request #1 from Berted/branch-A-Assertions

Add assertions on string parsing methods

Berted committed:
Adds Archive command

Berted committed:
Merge branch 'branch-C-Archive'

Berted committed:
GUI: Modifies anchor point for more responsive design

Berted committed:
Adds basic user guide

Berted committed:
README: Adds Quickstart, Summary, and styling

Berted committed:
README: Adds Troubleshooting, Changes Theme

Berted committed:
README: Modifies README.md

Berted committed:
Adds support to run in CLI mode
